wandb/sdk/launch/runner/kubernetes.py

```diff
@@ -0,0 +1,222 @@
+            job_metadata['labels'] = resource_args.get('job_labels')
+
+        if resource_args.get('backoff_limit'):
+            job_spec['backoffLimit'] = resource_args.get('backoff_limit')
+        if resource_args.get('job_completions'):
+            job_spec['completions'] = resource_args.get('job_completions')
+        if resource_args.get('parallelism'):
+            job_spec['parallelism'] = resource_args.get('parallelism')
+        if resource_args.get('suspend'):
+            job_spec['suspend'] = resource_args.get('suspend')
+
+        # only support container overrides for the single container case
+        if any(arg in resource_args for arg in ['container_name', 'resource_requests', 'resource_limits']):
+            raise LaunchError("Resource overrides not supported for multiple containers. Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+
+        containers[0]['name'] = resource_args.get("container_name", containers[0].get('name', 'launch'))
+        container_resources = containers[0].get('resources', {})
+        if resource_args.get('resource_requests'):
+            container_resources['requests'] = resource_args.get("resource_requests")
+        if resource_args.get('resource_limits'):
+            container_resources['limits'] = resource_args.get('resource_limits')
+        if container_resources:
+            containers[0]['resources'] = container_resources
+        # todo: args and env vars would be added here, need to figure out what kind of overrides we want
+
+        pod_spec['restartPolicy'] = resource_args.get('pod_restart_policy', 'Never')
+        if resource_args.get('pod_preemption_policy'):
+            pod_spec['preemptionPolicy'] = resource_args.get("pod_preemption_policy")
+        if resource_args.get('node_name'):
+            pod_spec['nodeName'] = resource_args.get('node_name')
+        if resource_args.get('node_selectors'):
+            pod_spec['nodeSelectors'] = resource_args.get('node_selectors')
+
+        validate_docker_installation()
+        entry_point = launch_project.get_single_entry_point()
+        docker_args: Dict[str, Any] = self.backend_config[PROJECT_DOCKER_ARGS]
+        if docker_args and list(docker_args) != ['docker_image']:
+            wandb.termwarn(
+                "Docker args are not supported for Kubernetes. Not using docker args"
+            )
+
+        image = resource_args.get("image")
+        if image:
+            if launch_project.docker_image and image != launch_project.docker_image:
+                raise LaunchError("Conflicting Docker images specified in launch and resource args.")
+            if len(containers) > 1:
+                raise LaunchError("Multiple container configurations should be specified in a yaml file supplied via job_spec.")
+        user_provided_image = image or launch_project.docker_image
+        if user_provided_image:
+            pull_docker_image(user_provided_image)
```

Comment:
Why are we pulling an image here?  k8s will pull the image when it needs to run it...

Reply:
yup removing these (also from aws/gcp)